Gameobject rotation (from mangos/MaNGOS@2bcbc0f)

ariel- opened the pull request:
Ported functions and merged with newer core.

ariel- committed:
Gameobject rotation (from mangos/MaNGOS@2bcbc0f)

Goatform commented:
nice :+1:

Naios commented:
:+1:

Aokromes commented:
Anyone can test it and move around world searching for incorrectly spawned doors or other things?
Imho better fix db before push this taking data from UDB.

NorthStrider commented:
currently compiling and testing. gonna give feedback asap im done with some tests

NorthStrider commented:
well it appears that all doors are facing into 1 direction:
i rechecked the sniffs where the db spawns for shadow fang keep were comming from but it seems sth is missing (tdb spawns ofc, sfk was spawned by me back in closed tdb 434 times)

![wowscrnshot_021415_185301](https://cloud.githubusercontent.com/assets/1472908/6200488/22f9666c-b47b-11e4-88ff-9aacebf1a866.jpg)
![wowscrnshot_021415_184131](https://cloud.githubusercontent.com/assets/1472908/6200489/2f1a57da-b47b-11e4-829a-289317f5ac5b.jpg)
![wowscrnshot_021415_184151](https://cloud.githubusercontent.com/assets/1472908/6200490/2fcaedc0-b47b-11e4-966e-15d0719f5021.jpg)

ariel- commented:
Many GOs are fine, some slanted examples here:

![wowscrnshot_021515_011054](https://cloud.githubusercontent.com/assets/9031719/6202067/8490de8e-b4b0-11e4-9da0-54d609b7593a.jpg)
![wowscrnshot_021515_011258](https://cloud.githubusercontent.com/assets/9031719/6202070/8527f166-b4b0-11e4-9e0f-5fc7d64b12ee.jpg)
![wowscrnshot_021515_011338](https://cloud.githubusercontent.com/assets/9031719/6202069/85161540-b4b0-11e4-9049-4e4b069e97c3.jpg)
![wowscrnshot_021515_011440](https://cloud.githubusercontent.com/assets/9031719/6202071/852fa370-b4b0-11e4-832c-4526e27d7631.jpg)
![wowscrnshot_021515_011453](https://cloud.githubusercontent.com/assets/9031719/6202072/853149d2-b4b0-11e4-9c7d-424812e88148.jpg)
![Uploading WoWScrnShot_021515_011553.jpg . . .]()
![wowscrnshot_021515_011702](https://cloud.githubusercontent.com/assets/9031719/6202068/84d326ae-b4b0-11e4-9509-9812c1dd2c9b.jpg)

Golrag commented:
Maybe the SQL in the comment from the mangos commit can help
https://github.com/mangos/MaNGOS/commit/2bcbc0fbfe63fab389e4f4f875ac215852bff8b6
```sql
UPDATE gameobject SET rotation0 = 0, rotation1 = 0, rotation2 = SIN(orientation * 0.5), rotation3 = COS(orientation * 0.5);
```
Not sure if this is Blizzlike :/

Aokromes commented:
We have a problem, we have at least some GOs with proper rotation, we need to compare go id + x + y + z from TDB GOs to UDB GOs.

jackpoz commented on src/server/scripts/Commands/cs_gobject.cpp:
```diff
@@ -416,25 +416,37 @@ class gobject_commandscript : public CommandScript
+            if (orientation = strtok(NULL, " "))
+                ox = (float)atof(orientation);
+            else
+                ox = 0;
```
please use 0.f to clearly state the type of the number

CreshandKesh commented:
Fixes Dalaran, Data Taken from UDB
```sql
UPDATE gameobject SET rotation0=-0.010067,  rotation1=-0.0192642, rotation2=-0.54460,  rotation3=0.838411  WHERE guid=57794;
UPDATE gameobject SET rotation0=-0.0478525, rotation1=-0.0228243, rotation2=-0.90131,  rotation3=0.429906  WHERE guid=57849;
UPDATE gameobject SET rotation2=0.143493,   rotation3=0.989651  WHERE guid=60113;
UPDATE gameobject SET rotation2=0.25038,    rotation3=0.968148  WHERE guid=60120;
UPDATE gameobject SET rotation2=0.309016,   rotation3=0.951057  WHERE guid=60227;
UPDATE gameobject SET rotation2=0.426569,   rotation3=0.904455  WHERE guid=60244;
UPDATE gameobject SET rotation2=-0.902585,  rotation3=0.430512  WHERE guid=60253;
UPDATE gameobject SET rotation2=0.426569,   rotation3=0.904455  WHERE guid=60258;
UPDATE gameobject SET rotation2=0.945518,   rotation3=0.325568  WHERE guid=60297;
UPDATE gameobject SET rotation2=0.434444,   rotation3=0.900699  WHERE guid=60323;
UPDATE gameobject SET rotation2=0.450098,   rotation3=0.892979  WHERE guid=61476;
UPDATE gameobject SET rotation2=0.588305,   rotation3=0.808639  WHERE guid=56109;
UPDATE gameobject SET rotation2=-0.90258,   rotation3=0.430512  WHERE guid=57802;
UPDATE gameobject SET rotation2=-0.78801,   rotation3=0.615662  WHERE guid=57892;
UPDATE gameobject SET rotation2=0.999391,   rotation3=0.0349061 WHERE guid=57913;
UPDATE gameobject SET rotation2=-0.63607,   rotation3=0.771625  WHERE guid=57960;
UPDATE gameobject SET rotation2=0.999391,   rotation3=0.0349061 WHERE guid=58082;
UPDATE gameobject SET rotation2=0.130525,   rotation3=0.991445  WHERE guid=60303;
UPDATE gameobject SET rotation2=-0.808639,  rotation3=0.588306  WHERE guid=56101;
UPDATE gameobject SET rotation2=-0.824125,  rotation3=0.566408  WHERE guid=61486;
UPDATE gameobject SET rotation2=0.566406,   rotation3=0.824126  WHERE guid=61497;
UPDATE gameobject SET rotation2=0.430511,   rotation3=0.902586  WHERE guid=61606;
```

jackpoz commented on src/server/scripts/Commands/cs_gobject.cpp:
```diff
@@ -416,25 +416,37 @@ class gobject_commandscript : public CommandScript
-        object->UpdateRotationFields();
+        // try extract two opt-parameters
+        orientation = strtok(NULL, " ");
+        if (orientation)
```
https://travis-ci.org/TrinityCore/TrinityCore/builds/50880616#L1045
>variable 'ox' is used uninitialized whenever 'if' condition is false

was this part tested ingame (or even debugged) ?

ariel- committed:
Fix var ox sometimes not initializated (@jackpoz), small optimization for quat packing helper struct.

Aokromes commented:
NorthStrider / Ariel- can you check dalaran with this patch after apply https://github.com/TrinityCore/TrinityCore/pull/14146#issuecomment-74441429 ?

ariel- commented:
Did a before and after run and it fixes those guids, guild banks are still circa 90 degrees into the wall tho

CreshandKesh commented:
Now I see, also missing several objects such as chairs and mailboxes, working on it. thanx @ariel-

CreshandKesh commented:
Fixes :
https://gist.github.com/CreshandKesh/c3dcf0f8106ac84ce82f

Aokromes commented:
IMHO the higher priority is fix doors on instances/entrances to instances before pushing this.

Aokromes commented:
```sql
select * from udb.gameobject as t where t.id in (select id from world.gameobject as u inner join world.gameobject_template as ugt on u.id = ugt.entry where ugt.type = 0 and u.id = t.id and u.position_x = t.position_x and  u.position_y = t.position_y and  u.position_z = t.position_z and (u.rotation0 != t.rotation0 or u.rotation1 != t.rotation1 or u.rotation2 != t.rotation2 or u.rotation3 != t.rotation3));

619 rows in set (15.98 sec)

Golrag commented:
What about this commit ?https://github.com/mangos/MaNGOS/commit/0fe88f35dfb2e4f9454d160cdad439e7acac326e

Naios commented:
@ariel- probably you should take a look at the current head of mangos gameobject rotation to include their whole work.

ariel- commented:
True that, probably trying to port HEAD is a better idea, I'll keep updating this

ariel- committed:
Pick some more MaNGOS commits, little rewording:

mangos/MaNGOS@0fe88f35dfb: [11531] Normalize gameobject's quaternion, thanks to zergtmn for pointing
mangos/MaNGOS@060dfb791b: [11667] Implement transport path rotation
mangos/MaNGOS@565f52c6c1: [11806] A bit gameobject code refactoring

ariel- commented:
Transports issue: MaNGOS maintained a table in DB with custom rotation values for transports, as of now, applying this wrecks the deeprun tram eg (goes the other way around)

jackpoz commented on src/server/game/Entities/GameObject/GameObject.h:
```diff
@@ -32,6 +32,18 @@ class Transport;
 #define MAX_GAMEOBJECT_QUEST_ITEMS 6
 
+// small wrapper to avoid clutter
+struct QuatData
```
would it make sense to use G3D Quaternion ?

armano2 replied:
yes G3D::Quat should be used here

ariel- commented:
Idk, that's how it was on mangos, maybe to avoid extra overhead? since they don't need all the extra operations/members a full fledged G3D quat has to offer, they only need the 4 coords.

jackpoz commented:
G3D::Quat is used at https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R1856 and at https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R1879 tho, it would be nice to be consistent all around the code

Arsul commented:
Hope this helps

Copied udb.gameobject and tdb.gameobject, then used the following query, which checks for same identity (guid, id, x, y, z) but different rotation values. Should I include Orientation into the Join too?


select t.guid, t.id, t.map, t.spawnmask, t.phasemask, t.position_x, t.position_y, t.position_z, t.orientation, u.rotation0, u.rotation1, u.rotation2, u.rotation3, t.spawntimesecs, t.animprogress, t.state from gameobject as t 
JOIN gameobject.gameobjectcmangos as u on u.guid=t.guid and u.id=t.id and u.position_x=t.position_x and u.position_y=t.position_y and u.position_z=t.position_z
WHERE ((u.rotation3 - t.rotation3 > 0.01) OR (u.rotation2 - t.rotation2 > 0.01) OR (u.rotation1 - t.rotation1 > 0.01) OR (u.rotation0 - t.rotation0 > 0.01) OR (u.rotation3 - t.rotation3 < -0.01) OR (u.rotation2 - t.rotation2 < -0.01) OR (u.rotation1 - t.rotation1 < -0.01) OR (u.rotation0 - t.rotation0 < -0.01))

Output where 47637 Lines with different values. Some Copy&Paste Magic and the following SQL File was the result:

http://www.file-upload.net/download-10515873/Gameobjectrotation.sql.html

Any suggestions?

DDuarte commented:
Deleting and inserting a ton of gameobjects where a update would be enough might not be a great idea.

http://software.dell.com/products/toad-for-mysql/ can be used to compare two DBs and generate UPDATE queries.

ariel- commented:
Maybe something like [this](https://github.com/TrinityCore/TrinityCore/issues/13750#issuecomment-73254334) can be used (I create a temporary table with all the updates, populate it only with the columns that are changed, then update and drop)

jackpoz commented:
any progress on the sql part of this pr ?

ariel- commented:
Is it ok to automatically asume that UDB has true rotation values and TDB does not? or will I need some sniffs?

Nyeriah commented:
It has been a decade but doesn't UDB have VerifiedBuild too? You could check that, though checking sniffs would actually be better

ariel- committed:
Code refactor

ariel- commented:
Updated, still missing some GO's (specially those that were nearby but not exactly on the same position in each DB)

Aokromes commented:
@ariel- :) and jump into irc xd
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

tkrokli commented:
The PR branch `ariel-:gobrot` is 520 commits behind `TrinityCore:3.3.5` - needs to be updated.

ariel- committed:
Port mangos/MaNGOS@6874951

[11807] Add gameobject_addon table

table holds additional per-guid gameobject data.
currently it contains path rotation info that required for some elevators and transports

msoky commented:
can be merged?

Rushor commented:
Updated, still missing some GO's (specially those that were nearby but not exactly on the same position in each DB)

this needs to be done first, else we will have thousands of userreports concerning wrong go positions

jackpoz commented:
I heard this change should have a PR on WPP too otherwise it will be a mess with 2 different structures

Shauren commented:
Instead of making this large sql file, why didn't you just write 3 queries like this
```
UPDATE gameobject_addon ga JOIN gameobject g ON ga.guid = g.guid SET ga.path_rotation0 = g.rotation0, ga.path_rotation1 = g.rotation1, ga.path_rotation2 = g.rotation2, ga.path_rotation3 = g.rotation3;
INSERT IGNORE INTO gameobject_addon SELECT guid, rotation0, rotation1, rotation2, rotation3, 0, 0 FROM gameobject WHERE rotation0 != 0 OR rotation1 != 0 OR rotation2 != 0 OR rotation3 != 1;
UPDATE gameobject SET rotation0 = 0, rotation1 = 0; -- existing core calculation always ended up with this
--- 3rd query calculates unpacked rotation2 and rotation3 from orientation like this https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Entities/GameObject/GameObject.cpp#L1958 ---
```
And then finally after that, port any mangos db rows that have rotation0 != 0 or rotation1 != 0 - there are only 565 of them in your sql, compared to the current thousands

ariel- commented:
I don't entirely like it, it encourages for bad future spawning of GOs with this code in place.

https://github.com/TrinityCore/TrinityCore/pull/14146/files#diff-fe5c8ebb7ad8e663836b00f39eb524d6R221

PARENT_ROTATION field is only for transport type GameObject; it shouldn't be used to hack and send to client existing bad data.

Shauren commented:
What I meant is porting that method into the one-time sql for existing spawns to minimize potential errors (and obviously remove it from code)

ariel- commented:
I'm updating this. However I've run into a problem.

You see, rotation fields and GAMEOBJECT_PARENTROTATION are two different things, in sniffs only some gameobjects have GAMEOBJECT_PARENTROTATION sent but all of them have rotations.

The above described behaviour makes it similar to creature_addon that sets fields for some (but not all) spawns, which is exactly what gameobject_addon is for.

There are few 'elegant' options besides re-spawning the gameobject DB, I tried with yours but it'll just ruin the few 'good' spawns TC has (sending them to addon is not an option, or would require ANOTHER set of 4 fields, because as stated before GAMEOBJECT_PARENTROTATION != packed rotation)

So, here comes the "intermediate" approach: Spawn whitelisting
Add a flag in gameobject_addon table: with this flag set, gameobject will use rotation fields from db, if not set or 0, gameobject will be sent to client with current (or legacy) rotation values (ie calculated from orientation)

ariel- committed:
Rebasing, some code tidying and renames, logic fail fetching gameobjectaddon fixed (this was started long before map local guids)

Shauren commented:
I still have a feeling you are not understanding what I mean
* Moving current gameobject.rotation fields into gameobject_addon with no alteration is safe - old code loads gameobject.rotation into PARENTROTATION, new code loads gameobject_addon into PARENTROTATION
* Old code has no source for storing packed rotation, instead it calculates it from orientation using this code
```
    static double const atan_pow = atan(pow(2.0f, -20.0f));

    double f_rot1 = std::sin(GetOrientation() / 2.0f);
    double f_rot2 = std::cos(GetOrientation() / 2.0f);

    int64 i_rot1 = int64(f_rot1 / atan_pow *(f_rot2 >= 0 ? 1.0f : -1.0f));
    int64 rotation = (i_rot1 << 43 >> 43) & 0x00000000001FFFFF;

    m_rotation = rotation;
```
* New code will load packed rotation from gameobject.rotation fields - and this is what we need to fill - using the method above (adapted for sql) we calculate gameobject.rotation from orientation
Having this done, all objects should have 100% the same values sent to clients as before

After this is done, the db should have all gameobject.rotation0 and gameobject.rotation1 fields set to 0 because there was no support for rotating around X and Y axes before (spawns that do need it can be ported from mangos db)

Warpten commented:
I think it would be best to just store orientation as a quat directly in db; after all even floats for rotations don't mean anything to regular humans and you basically save the hassle of dealing with arbitrary precision

ariel- commented:
What would be a good mysql data type for storing quats directly then?

Shauren commented:
4x float, ignore warptens random math stuff

ariel- committed:
SQL updated, renamed fields to parent_rotationN

ariel- commented:
That should do it, trouble spots I checked are now gone.

Shauren commented on src/server/game/Entities/GameObject/GameObject.cpp:
```diff
@@ -1899,34 +1893,46 @@ std::string const & GameObject::GetNameForLocaleIdx(LocaleConstant loc_idx) cons
+{
+    G3D::Quat rotation;
+    // Temporary solution for gameobjects that have no rotation data in DB:
+    if (G3D::fuzzyEq(rot.z, 0.f) && G3D::fuzzyEq(rot.w, 0.f))
```
This is not needed thanks to the query doing the calculation
